### ISSUE TITLE
LXC init does not pass template param to lxc-create

### ISSUE THREAD

**Issue by martinhoefling** [saltstack/salt#27662]
Creating a new lxc instance via salt-cloud fails, probably because the `-t templatename` parameter is missing.

Output:
```
root@debian-jessie:/home/vagrant# salt-cloud -p ubuntu-lxc test
[INFO    ] salt-cloud starting
themaster:
    True
[INFO    ] Runner completed: 20151004103030737616
themaster:
    True
[INFO    ] Runner completed: 20151004103031273046
themaster:
    ----------
    frozen:
        ----------
    running:
        ----------
    stopped:
        ----------
[INFO    ] Runner completed: 20151004103031983817
[INFO    ] Searching for LXC Hosts
[INFO    ] Creating container(s) '['test']' on host 'themaster'
event:
    ----------
    message:
        ----------
        comment:
        done:
        errors:
            ----------
            themaster:
                ----------
                test:
                    |_
                      ----------
                      changes:
                          ----------
                      comment:
                          Container could not be created with cmd 'lxc-create -n test -f /tmp/tmpxgBXi5 -B dir': lxc_container: Error creating container test
                      name:
                          test
        ping_status:
            False
        result:
            False
        themaster:
suffix:
    progress
comment:
done:
errors:
    ----------
    themaster:
        ----------
        test:
            |_
              ----------
              changes:
                  ----------
              comment:
                  Container could not be created with cmd 'lxc-create -n test -f /tmp/tmpxgBXi5 -B dir': lxc_container: Error creating container test
              name:
                  test
ping_status:
    False
result:
    False
themaster:
[INFO    ] Runner completed: 20151004103032533633
test:
    ----------
    Error:
        Error while creating test,
    changes:
        ----------
    comment:
    name:
        test
    result:
        False
    runner_return:
        ----------
        comment:
        done:
        errors:
            ----------
            themaster:
                ----------
                test:
                    |_
                      ----------
                      changes:
                          ----------
                      comment:
                          Container could not be created with cmd 'lxc-create -n test -f /tmp/tmpxgBXi5 -B dir': lxc_container: Error creating container test
                      name:
                          test
        ping_status:
            False
        result:
            False
        themaster:
```

Configuration and a vagrant setup to reproduce the issue can be found in this repo:
https://github.com/martinhoefling/molten

* The configuration (master / cloud profiles, ...) are in [./vagrant/salt](https://github.com/martinhoefling/molten/tree/master/vagrant/salt)
* Run vagrant up dev-cloud in the repo (this creates a single instance with salt master / minion / lxc installed and configured)
* `vagrant ssh dev-cloud`, `sudo bash`, `salt-cloud -p ubuntu-lxc test`

If I run `lxc-create -t ubuntu -n test -B dir` everything works fine. Without `-t ubuntu`, I get the same output as in the salt init job.

**Comment by jtand** [comment]
@martinhoefling Thanks for the bug report!

**Comment by pass-by-value** [comment]
You are right @martinhoefling we do need the `image` key. There are some other keys required as well

Here's my cloud profile (notice the addition of `image` and `script_args`)

```yaml
ubuntu-lxc:
  provider: dev-lxc
  lxc_profile: lxc-ubuntu-default
  image: ubuntu
  script_args: -P
  network_profile: ubuntu
  minion:
    master: localhost
```

I added the following [network profile](https://docs.saltstack.com/en/latest/topics/tutorials/lxc.html#network-profiles) so the container can resolve addresses

```yaml
lxc.network_profile:
  ubuntu:
    eth0:
      link: virbr0
```

I got the name `vibr0` by doing

```shell
[root@li1258-28 ~]# brctl show
bridge name     bridge id               STP enabled     interfaces
virbr0          8000.525400f732bb       yes             vethFFLHGM
                                                        virbr0-nic
```

fyi: Distribution that I'm using

```shell
[root@li1258-28 ~]# python -c 'import platform; print platform.platform()'
Linux-4.1.5-x86_64-linode61-x86_64-with-centos-7.1.1503-Core
```

I'm on `2015.8`

Once I configure the above cloud profile and the network profile I get `Container 'test1' successfully initialized`.

Could you please try with these settings? Do let me know how it goes!

**Comment by martinhoefling** [comment]
Hmm, my profile now looks like this:

```
debian-lxc:
  provider: lxc-dev
  lxc_profile:
    template: download
    options:
      dist: debian
      release: jessie
      arch: amd64
  image: debian
  script_args: -P
  network_profile:
    eth0:
      link: virbr0
      type: veth
      flags: up
      ipv4: 192.168.122.100/24
      ipv4.gateway: 192.168.122.1
    eth1:
      link: virbr1
      type: veth
      flags: up
      ipv4: 192.168.100.100/24
  minion:
    master: 192.168.50.1
```

Now it seems as if the options are not supported by lxc:

```
muh2:
    ----------
    Error:
        Error while creating muh2,
    changes:
        ----------
    comment:
    name:
        muh2
    result:
        False
    runner_return:
        ----------
        comment:
        debian:
        done:
        errors:
            ----------
            debian:
                ----------
                muh2:
                    |_
                      ----------
                      changes:
                          ----------
                      comment:
                          Container could not be created with cmd 'lxc-create -n muh2 -f /tmp/tmpkgSR26 -t debian -B dir -- --release jessie --arch amd64 --dist debian': getopt: unrecognized option '--dist'
                          lxc_container: container creation template for muh2 failed
                          lxc_container: Error creating container muh2
                      name:
                          muh2
        ping_status:
            False
        result:
            False
```

However, this should be supported by lxc as documented here: https://github.com/lxc/lxc/blob/master/templates/lxc-download.in

Anyhow, this problem probably is not related to salt.

**Comment by pass-by-value** [comment]
@martinhoefling 

I applied the following profile and got `Container 'container27' successfully initialized`

```yaml
debian-lxc:
  provider: dev-lxc
  lxc_profile:
    template: download
    options:
      release: jessie
      arch: amd64
  image: debian
  script_args: -P
  network_profile: mynetwork
  minion:
    master: localhost
```

I'd suggest removing `dist` from under `options`. On my system the following options are supported

```shell
/usr/share/lxc/templates/lxc-debian -h|--help -p|--path=<path> [-a|--arch] [-r|--release=<release>] [-c|--clean]
```

```shell
root@localhost:~/salt# dpkg -l | grep lxc
ii  lxc                            1:1.0.6-6+deb8u1             amd64        Linux Containers userspace tools
```

**Comment by martinhoefling** [comment]
OK, so I got a bit farther with your hints:

```
[INFO    ] Runner completed: 20151031143511471821
[DEBUG   ] LazyLoaded nested.output
muh:
    ----------
    Error:
        Error while creating muh,
    changes:
        ----------
    comment:
    name:
        muh
    result:
        False
    runner_return:
        ----------
        comment:
        done:
        errors:
            ----------
            themaster:
                ----------
                muh:
                    |_
                      ----------
                      changes:
                          ----------
                          init:
                              |_
                                ----------
                                create:
                                    Container created
                              |_
                                ----------
                                config:
                                    Container configuration updated
                      comment:
                          Unable to stop container: Error changing state for container 'muh' using command 'lxc-start -f /var/lib/lxc/muh/config -d -n muh': lxc-start: The container failed to start.
                          lxc-start: To get more details, run the container in foreground mode.
                          lxc-start: Additional information can be obtained by setting the --logfile and --logpriority options.
                      name:
                          muh
        ping_status:
            False
        result:
            False
        themaster:
```

When I start the created container manually:
```
root@debian-jessie:/home/vagrant# lxc-start --name muh
lxc-start: failed to find gateway addresses
lxc-start: failed to spawn 'muh'
lxc-start: The container failed to start.
lxc-start: Additional information can be obtained by setting the --logfile and --logpriority options.
```

lxc container config:
```
# Template used to create this container: /usr/share/lxc/templates/lxc-ubuntu
# Parameters passed to the template: --release trusty --arch amd64
# For additional config options, please look at lxc.container.conf(5)

# Common configuration
lxc.include = /usr/share/lxc/config/ubuntu.common.conf

# Container specific configuration
lxc.rootfs = /var/lib/lxc/muh/rootfs
lxc.mount = /var/lib/lxc/muh/fstab
lxc.utsname = muh
lxc.arch = amd64

# Network configuration
lxc.start.auto = 1
lxc.network.type = veth
lxc.network.name = eth0
lxc.network.flags = up
lxc.network.link = br0
lxc.network.hwaddr = AC:DE:48:26:95:4C
lxc.network.ipv4.gateway = auto
```

My lxc related config in master.conf:
```
lxc.container_profile:
  lxc-ubuntu-default:
    template: ubuntu

lxc.network_profile:
  lxc-default:
    eth0:
      link: virbr0
      type: veth
      flags: up
```

So the virbr0 seems to be ignored during create.

**Comment by pass-by-value** [comment]
Hi @martinhoefling 

Thanks for working with me on this.

My cloud profile looks like

```yaml
ubuntu-lxc:
  provider: dev-lxc
  lxc_profile:
    template: download
    options:
      release: trusty
      arch: amd64
  image: ubuntu
  seed: False
  network_profile: ubuntu
  minion:
    master: localhost
```

And I have the following network profile defined in my minion config

```yaml
lxc.network_profile:
  ubuntu:
    eth0:
      link: virbr0
```

That lets me create a container with the following config

```yaml
# Template used to create this container: /usr/share/lxc/templates/lxc-ubuntu
# Parameters passed to the template: --release trusty --arch amd64
# For additional config options, please look at lxc.container.conf(5)

# Common configuration
lxc.include = /usr/share/lxc/config/ubuntu.common.conf

# Container specific configuration
lxc.rootfs = /var/lib/lxc/some_new_container27/rootfs
lxc.mount = /var/lib/lxc/some_new_container27/fstab
lxc.utsname = some_new_container27
lxc.arch = amd64

# Network configuration
lxc.start.auto = 1
lxc.network.type = veth
lxc.network.name = eth0
lxc.network.flags = up
lxc.network.link = virbr0
lxc.network.hwaddr = AC:DE:48:47:BB:CF
```

I'm wondering if your network profile gets overridden somewhere?

I checked the profile settings like below

```shell
root@localhost:~/salt# salt localminion lxc.get_network_profile ubuntu
localminion:
    ----------
    eth0:
        ----------
        link:
            virbr0
```

**Comment by martinhoefling** [comment]
@pass-by-value 

So I now also have my network profile in the minion not the master config:
```
root@debian-jessie:/etc/salt/cloud.profiles.d# salt themaster lxc.get_network_profile lxc-default
themaster:
    ----------
    eth0:
        ----------
        link:
            virbr0
```

The result is however still the same: I have an lxc.network.ipv4.gateway = auto in my lxc container config.

Also tried your profile and network config with the same result.

What is your host system and salt version? I am using vanilla debian stable and salt deb packages:
```
root@debian-jessie:/etc/salt/cloud.profiles.d# salt --versions-report
Salt Version:
           Salt: 2015.8.0

Dependency Versions:
         Jinja2: 2.7.3
       M2Crypto: Not Installed
           Mako: 1.0.0
         PyYAML: 3.11
          PyZMQ: 14.4.0
         Python: 2.7.9 (default, Mar  1 2015, 12:57:24)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: 0.8.6
       cherrypy: 3.5.0
       dateutil: 2.2
          gitdb: 0.5.4
      gitpython: Not Installed
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: Not Installed
      pycparser: 2.10
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: 0.8.2
        timelib: Not Installed

System Versions:
           dist: debian 8.1
        machine: x86_64
        release: 3.16.0-4-amd64
         system: debian 8.1
```

**Comment by pass-by-value** [comment]
Debian 8.1 is what I'm using as well.

I installed salt via the bootstrap (but I  have updated my `PATH` and `PYTHONPATH` variables so I can run from a source checkout).

root@localhost:~/salt# salt --versions-report
Salt Version:
           Salt: 2015.8.1-157-g93a9306

Dependency Versions:
         Jinja2: 2.7.3
       M2Crypto: 0.21.1
           Mako: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.4.0
         Python: 2.7.9 (default, Mar  1 2015, 12:57:24)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: 0.8.6
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: Not Installed
      pycparser: 2.10
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: 0.3.8
          smmap: Not Installed
        timelib: Not Installed

System Versions:
           dist: debian 8.1
        machine: x86_64
        release: 4.1.5-x86_64-linode61
         system: debian 8.1

root@localhost:~/salt#

Do you see any errors in the salt minion logs? If possible I'd like to see a gist of your salt minion logs when you run the salt-cloud command.

**Comment by martinhoefling** [comment]
I think line https://github.com/saltstack/salt/blob/develop/salt/modules/lxc.py#L868 is wrong. It should also exclude 1.0.6 (the default version in jessie) from adding a gateway = auto. If I remove that line from the config, the container can start.

**Comment by martinhoefling** [comment]
Next stage: The bootstrap fails

```
[INFO    ] Running sh /sbin/d1de3c6c3a249617d70fa096f657bbef_bootstrap.sh -P -c /var/tmp/.c_d1de3c6c3a249617d70fa096f657bbef in LXC container 'cont3'
[DEBUG   ] Child Forked! PID: 16858  STDOUT_FD: 27  STDERR_FD: 29
[DEBUG   ] Terminal Command: /bin/sh -c lxc-attach --clear-env --set-var PATH=/bin:/usr/bin:/sbin:/usr/sbin:/opt/bin:/usr/local/bin:/usr/local/sbin  -n cont3 -- sh /sbin/d1de3c6c3a249617d70fa096f657bbef_bootstrap.sh -P -c /var/tmp/.c_d1de3c6c3a249617d70fa096f657bbef
[INFO    ]  *  INFO: sh /sbin/d1de3c6c3a249617d70fa096f657bbef_bootstrap.sh -- Version 2015.11.04

 *  INFO: System Information:
 *  INFO:   CPU:          GenuineIntel
 *  INFO:   CPU Arch:     x86_64
 *  INFO:   OS Name:      Linux
 *  INFO:   OS Version:   3.16.0-4-amd64
 *  INFO:   Distribution: Ubuntu 14.04

 *  INFO: Installing minion
 *  INFO: Found function install_ubuntu_stable_deps
 *  INFO: Found function config_salt
 *  INFO: Found function install_ubuntu_stable
 *  INFO: Found function install_ubuntu_restart_daemons
 *  INFO: Found function daemons_running
 *  INFO: Found function install_ubuntu_check_services
 *  INFO: Running install_ubuntu_stable_deps()
Err http://security.ubuntu.com trusty-security InRelease
  
Err http://security.ubuntu.com trusty-security Release.gpg
  Temporary failure resolving 'security.ubuntu.com'
Err http://archive.ubuntu.com trusty InRelease
  
Err http://archive.ubuntu.com trusty-updates InRelease
  
Err http://archive.ubuntu.com trusty Release.gpg
  Te
[INFO    ] mporary failure resolving 'archive.ubuntu.com'
Err http://archive.ubuntu.com trusty-updates Release.gpg
  Temporary failure resolving 'archive.ubuntu.com'
Reading package lists...
[INFO    ] W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty/InRelease  

W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty-updates/InRelease  

W: Failed to fetch http://security.ubuntu.com/ubuntu/dists/trusty-security/InRelease  

W: Failed to fetch http://security.ubuntu.com/ubuntu/dists/trusty-security/Release.gpg  Temporary failure resolving 'security.ubuntu.com'

W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty/Release.gpg  Temporary failure resolving 'archive.ubuntu.com'

W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/trusty-updates/Release.gpg  Temporary failure resolving 'archive.ubuntu.com'

W: Some index files failed to download. They have been ignored, or old ones used instead.
Reading package lists...
Building dependency tree...
Reading state information...
[INFO    ] The following NEW packages will be installed:
  debian-archive-keyring
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 34.4 kB of archives.
After this operation, 84.0 kB of additional disk space will be used.
Err http://archive.ubuntu.com/ubuntu/ trusty/universe debian-archive-keyring all 2012.4
  Temporary failure resolving 'archive.ubuntu.com'
E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/universe/d/debian-archive-keyring/debian-archive-keyring_2012.4_all.deb  Temporary failure resolving 'archive.ubuntu.com'

E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
Reading package lists...
Building dependency tree...
Reading state information...
The following extra packages will be installed:
  ca-certificates gir1.2-glib-2.0 iso-codes libasn1-8-heimdal libcurl3-gnutls
  libdbus-glib-1-2 libgirepository-1.0-1 libglib2.0-0 libglib2.0-data
  libgssapi3-heimdal libhcrypto4-heimdal libheimbase1-heimdal
  libheimntlm0-heimdal lib
[INFO    ] hx509-5-heimdal libidn11 libkrb5-26-heimdal
  libldap-2.4-2 libroken18-heimdal librtmp0 libsasl2-2 libsasl2-modules
  libsasl2-modules-db libwind0-heimdal libxml2 openssl python-apt-common
  python3-apt python3-dbus python3-gi python3-pycurl
  python3-software-properties sgml-base shared-mime-info unattended-upgrades
  xml-core xz-utils
Suggested packages:
  isoquery libsasl2-modules-otp libsasl2-modules-ldap libsasl2-modules-sql
  libsasl2-modules-gssapi-mit libsasl2-modules-gssapi-heimdal python3-apt-dbg
  python-apt-doc python-dbus-doc python3-dbus-dbg libcurl4-gnutls-dev
  python3-pycurl-dbg sgml-base-doc bsd-mailx mail-transport-agent debhelper
The following NEW packages will be installed:
  ca-certificates gir1.2-glib-2.0 iso-codes libasn1-8-heimdal libcurl3-gnutls
  libdbus-glib-1-2 libgirepository-1.0-1 libglib2.0-0 libglib2.0-data
  libgssapi3-heimdal libhcrypto4-heimdal libheimbase1-heimdal
  libheimntlm0-heimdal libhx509-5-heimdal libidn11 libkrb5-26-heimdal
  libldap-2.4-2 libroken
[INFO    ] 18-heimdal librtmp0 libsasl2-2 libsasl2-modules
  libsasl2-modules-db libwind0-heimdal libxml2 openssl python-apt-common
  python3-apt python3-dbus python3-gi python3-pycurl
  python3-software-properties sgml-base shared-mime-info
  software-properties-common unattended-upgrades xml-core xz-utils
0 upgraded, 37 newly installed, 0 to remove and 0 not upgraded.
Need to get 7173 kB of archives.
After this operation, 37.3 MB of additional disk space will be used.
WARNING: The following packages cannot be authenticated!
  libroken18-heimdal libasn1-8-heimdal libhcrypto4-heimdal
  libheimbase1-heimdal libwind0-heimdal libhx509-5-heimdal libkrb5-26-heimdal
  libheimntlm0-heimdal libgssapi3-heimdal libldap-2.4-2 libcurl3-gnutls
  libglib2.0-0 libxml2 openssl ca-certificates libgirepository-1.0-1
  gir1.2-glib-2.0 libglib2.0-data python-apt-common python3-apt python3-gi
  unattended-upgrades python3-software-properties software-properties-common
E: There are problems and -y was used without --force-yes
[INFO    ]  * ERROR: Failed to run install_ubuntu_stable_deps()!!!
[INFO    ] Executing command 'lxc-attach --clear-env --set-var PATH=/bin:/usr/bin:/sbin:/usr/sbin:/opt/bin:/usr/local/bin:/usr/local/sbin  -n cont3 -- sh -c \'if [ -f "/sbin/d1de3c6c3a249617d70fa096f657bbef_bootstrap.sh" ];then rm -f "/sbin/d1de3c6c3a249617d70fa096f657bbef_bootstrap.sh";fi\'' in directory '/root'
[DEBUG   ] Minion return retry timer set to 2 seconds (randomized)
[INFO    ] Returning information for job: 20151105201658997932
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/minion', 'themaster', 'tcp://127.0.0.1:4506', 'aes')
[DEBUG   ] Initializing new SAuth for ('/etc/salt/pki/minion', 'themaster', 'tcp://127.0.0.1:4506')
     
```

I can manually enter the running container via lxc-attach. Name resolution and internet connectivity works. Any ideas what could be wrong? I've updated the [repo](https://github.com/martinhoefling/molten) with the vagrant setup to reflect the most recent changes.

**Comment by pass-by-value** [comment]
@martinhoefling  Per your [comment](https://github.com/saltstack/salt/issues/27662#issuecomment-154172109) As far as the dist utils version goes I'm not sure if there is a specific reason why 1.0.7 is used there. I did a git annotate and it seems like @kiorky might have some information about it. Could you please let us know about the version @kiorky ?

**Comment by martinhoefling** [comment]
Bootstrapping manually via `salt 'themaster' lxc.bootstrap cont3` works. Seems to be a timing issue - salt doesn't wait until the container got an address via dhcp. So maybe adding a sleep before the bootstrapping process helps?

**Comment by pass-by-value** [comment]
@martinhoefling Per your [other comment](https://github.com/saltstack/salt/issues/27662#issuecomment-154180179), I just added a [PR](https://github.com/saltstack/salt/pull/28610) that lets you add a delay to bootstrap in the config, that should let you work around the address resolution errors. But there seems to be a [regression](https://github.com/saltstack/salt-bootstrap/issues/686) in the latest bootstrap that will prevent you from installing salt. The older version of bootstrap does not have this issue btw.

So in summary, there is a commit in the `2015.8` branch that takes care of the address resolution errors but the latest bootstrap on `Ubuntu 14.04` still has an issue.

**Comment by pass-by-value** [comment]
Yup that's a good point @martinhoefling I would suggest adding a delay using `bootstrap_delay`. But you will have to have the latest `2015.8` checked out for it since that lets you add the bootstrap delay in your cloud profile. Before that commit there was no way to set it (afaict) other than manually setting it in the [code](https://github.com/saltstack/salt/blob/develop/salt/modules/lxc.py#L3404).

**Comment by martinhoefling** [comment]
Just found out as well that there's a `bootstrap_delay=None` in the code. So it would be nice if this is exposed to the config so that it can be set for environments.
So as far as I can see, all issues are addressed except the one waiting for @kiorky's answer to version 1.0.7. Thx for your quick responses @pass-by-value. As soon as the setup is working, I'll update the vagrant setup since that would make a nice demo / playground for salt-cloud which doesn't require AWS or any other 'real' IaaS provider.

**Comment by basepi** [comment]
@pass-by-value is this resolved? Should I bump it to a more recent milestone?

**Comment by pass-by-value** [comment]
@basepi This is resolved, except for the distutils version (1.0.7). Does that still cause an issue @martinhoefling ?

**Comment by basepi** [comment]
Since part of this is resolved and it's a few months old, I'm going to go ahead and close it. If there are still issues we can either re-open this or create a new issue for anything remaining.
